### PR TITLE
Mamba monkey-patching + MoE any-2-any

### PR DESCRIPTION
**Pull request by Krystian-Krol**
# Description
Requires #358 first and rebase afterwards.
- Change naming for logging as one mamba block can now have multiple routers.
- Make token choice and expert choice any-to-any dimension.
- Make Mamba with MoE as linear projection.

[Neptune](https://app.neptune.ai/o/pmtest/org/llm-random/runs/details?viewId=9ac46a51-c84e-442d-963c-ac6a3a8af3f3&detailsTab=charts&shortId=LLMRANDOM-7220&dash=charts&type=run)

# Checklist
   - [x] Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - [x] Are all variables set with logical defaults that are backward compatible?
   - [x] I attached link to neptune if it was necessary
   - [x] Have you successfully executed the linter and tests?
   - [x] Are all functions concise and don't require splitting?
   - [x] Is there documentation for arguments and complex logic?
   - [x] Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - [x] Are all functions placed in appropriate files?
   - [x] Is the PR name meaningful and descriptive?
   - [x] Is PR description provided, or unnecessary?
  
Other questions:
   - [x] If I made changes to the requirements or anything related to the Singularity image I did generate new image and upload a new image to the clusters
   - [x] If this change is significant enough I notified all individuals working with this repository (@channel in slack channel)
   - [x] If there is a need for a second reviewer I requested additional review 

# Reviewer's checklist:
   - [ ] Does the Neptune link work and does it actually compare the situation before and after the PR?
   - [ ] Do I understand the code?
   - [ ] Do I think another reviewer is not needed or I commented that this PR need another reviewer?

### CODE REVIEW

File: research/conditional/utils/model_utils.py [hunk 1]
```diff
@@ -490,9 +518,81 @@ def get_mamba_layer(args):
...
+                init_type=args.init_type,
+            )
+            return mamba
+
+        return_fn = modified_out_mamba
+    elif args.mamba_mode == "conv_proj_moe":
+
+        def modified_in_mamba():
+            mamba = mamba_ssm.Mamba(d_model=args.dmodel, use_fast_path=False)
+            mamba.in_proj = MambaInProj(
```
[the remark this answers is not in the document]
Comment (sebastianjaszczur) [reply research/conditional/utils/model_utils.py:544]: Alright, my comment is invalid. I see Mamba.forward uses directly in_proj.weight, so you have to overwrite some matmuls. Actually clever.

**Review by sebastianjaszczur — COMMENTED** [review]
Looks good to me! At least correct and sensible; I wasn't thinking about engineering side at the moment, but it looks reasonably well. "fits into the repo"

File: lizrd/scripts/grid.py [hunk 2]
```diff
@@ -145,6 +145,7 @@
                 f"--gres=gpu:{setup_args['n_gpus']}",
                 "--partition=plgrid-gpu-a100",
                 f"--cpus-per-gpu={setup_args['cpus_per_gpu']}",
+                "--mem=96G",
```
Comment (ciebson123) [inline lizrd/scripts/grid.py:148]: Wyebać

File: research/conditional/utils/model_utils.py [hunk 3]
```diff
@@ -513,15 +514,75 @@ def get_ff_layer(args):
...
+    )
+
+
 def get_mamba_layer(args):
     import mamba_ssm
 
     if args.mamba_mode == "vanilla":
         return_fn = lambda: mamba_ssm.Mamba(
-            d_model=args.dmodel, expand=args.mamba_expansion
+            d_model=args.dmodel, expand=args.mamba_expansion, use_fast_path=False
```
Comment (ciebson123) [inline research/conditional/utils/model_utils.py:530]: Somewhat changes previous behaviour of vanilla mamba

**Review by ciebson123 — APPROVED** [review]

File: research/conditional/utils/model_utils.py (hunk 3 above)
Comment (Krystian-Krol) [reply research/conditional/utils/model_utils.py:530]: valid point but I think it's not significant, almost same steps/s afaik. I wanted to keep `use_fast_path=False` for consistency and modified Mamba requires it

**Pull request merged**